---
title: "Mouse and Rat glioblastoma models"
author: "Stefan Loska"
date: "March 14, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, libraries}
library(oligo)
library(Biobase)
library(AnnotationDbi)
```

# Introduction

# Downloading the data

```{r, eval=FALSE}
# gene homology data
download.file("ftp://ftp.ncbi.nih.gov/pub/HomoloGene/build68/homologene.data", "homologene.data")

# GEO Data
download.file("ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE12nnn/GSE12657/matrix/GSE12657_series_matrix.txt.gz",
              "GSE12657_series_matrix.txt.gz")

# Verhaak data
download.file("https://tcga-data.nci.nih.gov/docs/publications/gbm_exp/TCGA_unified_CORE_ClaNC840.txt", "Verhaak.tsv")
download.file("https://tcga-data.nci.nih.gov/docs/publications/gbm_exp/ClaNC840_centroids.xls", "ClaNC840_centroids.xls")

```

# Preparing microarray data

## Reading data

Read data from cel files and probe annotation file.

```{r }
read_cel <- function(data_dir, s_info, skip = NULL){
  # get sample meta data
  tab <- read.delim(paste(data_dir, s_info, sep = "/"), colClasses = c("character", "factor", "character"))
  rownames(tab) <- tab$filename
  # check sample_info vs. cel files
  if (! setequal(tab$filename, list.celfiles(data_dir))) stop("sample info file doesn't fit available cel files")

  # skip what is meant to skip
  tab <- tab[!rownames(tab) %in% skip,]

  # get microarray signal
  ab <- read.celfiles(filenames = paste(data_dir, tab$filename, sep = "/"), phenoData = AnnotatedDataFrame(tab))
  colnames(ab) <- tab$id
  ab
}
```


## Preprocessing

RMA

```{r }
rma_core <- function(ab){
  R <- oligo::rma(ab, target = "core")
  R
}
```


## Annotation

Get annotation and remove probes mapping to more than one gene.

```{r }
annotate <- function(R, annot){
  # choose columns for feature data
  fd <- select(annot, keys = rownames(R), keytype = "PROBEID", columns = c("ENTREZID", "SYMBOL", "GENENAME"))

  fd <- fd[!is.na(fd$ENTREZID),]

  freq <- table(fd$PROBEID)
  dambg <- names(freq)[freq == 1]
  fd <- fd[fd$PROBEID %in% dambg, ]
  rownames(fd) <- fd$PROBEID

  R <- R[rownames(fd)]
  fData(R) <- fd

  R
}

```

Pick up one probe for each gene. This will be the probe with the highest variance among samples.

```{r}
dedegen <- function(R){
  # keep only one probeset per gene; this will be the one with highest variance
  sel <- tapply(seq_along(fData(R)$ENTREZID), fData(R)$ENTREZID, c)

  sel <- tapply(seq_along(fData(R)$ENTREZID), fData(R)$ENTREZID, function(x){
    if (length(x) == 1) x else{
      vars <- apply(exprs(R)[x,], 1, var, na.rm = T)
      x[which.max(vars)]
    }
  })

  sel <- as.vector(sel)

  R <- R[sel,]
  R
}

```

Convert to human.

```{r}
to_human <- function(R, hom, taxid){
  # human to human just adds extra column in fData and changes rownames
  if (taxid == "9606") {
    fData(R)$HUMENTREZID <- fData(R)$ENTREZID
    rownames(R) <- fData(R)$HUMENTREZID
    return(R)
  }

  # restric to human and taxid
  hom <- hom[hom$tax %in% c("9606", taxid),]

  # remove disambiguities (and no pair), i.e. for each id need exactly one human and one taxid gene
  sel <- tapply(hom$tax, hom$id, function(x){
    length(x) == 2 & length(unique(x)) == 2
  })
  sel <- names(sel)[sel]
  hom <- hom[hom$id %in% sel,]

  # convert fData
  ids <- hom$id[match(fData(R)$ENTREZID, hom$entrez)]
  ens <- hom$entrez[match(ids, hom$id)]

  fData(R)$HUMENTREZID <- ens
  R <- R[!is.na(fData(R)$HUMENTREZID),]
  rownames(R) <- fData(R)$HUMENTREZID

  R
}

```

## Get expression

```{r}
get_exprs <- function(data_dir, s_info, skip = NULL, annot, hom, taxid){
  ab <- read_cel(data_dir, s_info, skip)
  R <- rma_core(ab)
  R <- annotate(R, annot)
  R <- dedegen(R)
  R <- to_human(R, hom, taxid)
  R
}

```


## Execute on data

```{r}
hom <- read.delim("homologene.data", header = F,
                  colClasses = c("character", "character", "character", "NULL", "NULL", "NULL"),
                  col.names = c("id", "tax", "entrez", "NULL", "NULL", "NULL"))

Rat <- get_exprs(data_dir = "Rat_data",
                 s_info = "sample_info.txt",
                 annot = ragene21sttranscriptcluster.db::ragene21sttranscriptcluster.db,
                 hom = hom,
                 taxid = "10116")

Mouse <- get_exprs(data_dir = "Mouse_data", 
          s_info = "sample_info.txt", 
          annot = mogene21sttranscriptcluster.db::mogene21sttranscriptcluster.db, 
          hom = hom, 
          taxid = "10090")

```


## PCA

```{r}
pca <- function(R, lbs = pData(R)$id){
  # svd
  e <- sweep(exprs(R), 1, rowMeans(exprs(R)))
  s <- svd(e)

  # plot
  pars <- par(no.readonly = T)
  on.exit(par(pars))
  par(pty = "s")
  # calcuklate variance percentage
  labs <- s$d^2
  labs <- round(labs[1:2]/sum(labs)*100, 0)
  labs <- paste(c("PC1: ", "PC2: "), labs, "%", sep = "")
  # draw
  plot(s$v[, 1:2], xlab = labs[1], ylab = labs[2], axes = F, pch = 16, col = as.numeric(pData(R)$class) + 1)
  box()
  abline(h = 0, v = 0, lty = 3)
  text(s$v[, 1], s$v[, 2], lbs, pos = 1)

  invisible(s)
}

```


Plotting PCA:

```{r}

pca(Rat)
pca(Mouse)

```


## Removing outliers

```{r}

Rat <- get_exprs(data_dir = "Rat_data", 
          s_info = "sample_info.txt", 
          skip = "C6_1.ga.cel", 
          annot = ragene21sttranscriptcluster.db::ragene21sttranscriptcluster.db, 
          hom = hom, 
          taxid = "10116")

```


PCA fter removing outliers:

```{r}
pca(Rat)
```


# Getting GEO data


```{r}
read_geo <- function(geo_file){
  tab <- readLines(geo_file)
  meta <- grepl("^!|^$", tab)

  pd <- grep("^!Sample_characteristics|^\"ID_REF\"", tab, value = T)
  pd <- gsub("\"", "", pd)
  pd <- strsplit(pd , "\t")
  pd <- do.call(cbind, pd)
  pd <- pd[-1, ]
  pd <- data.frame(class = factor(pd[, 1]), id = pd[, 2], stringsAsFactors = F)
  rownames(pd) <- pd$id

  ge <- tab[!meta]
  ge <- gsub("\"", "", ge)
  ge <- strsplit(ge , "\t")
  ge <- do.call(rbind, ge)
  colnames(ge) <- ge[1, ]
  rownames(ge) <- ge[, 1]
  ge <- ge[-1, -1]
  mode(ge) <- "numeric"
  ge <- log2(ge)

  G <- ExpressionSet(ge, AnnotatedDataFrame(pd))

  G
}

build_GEO <- function(geo_file, annot, hom, taxid){
  G <- read_geo(geo_file)
  G <- annotate(G, annot)
  G <- dedegen(G)
  G <- to_human(G, hom, taxid)
  G
}


G <- build_GEO("GSE12657_series_matrix.txt.gz", hgu95av2.db::hgu95av2.db, hom, "9606")

pca(G, lbs = pData(G)$class)
```


# Getting Verhaak data

## Expression data

```{r}
read_ver <- function(ver_file){
  # pheno data
  pd <- read.delim(ver_file, header = F, nrow = 2)
  pd <- t(pd[, -(1:2)])
  pd <- data.frame(class = (factor(pd[,2])), id = pd[,1], stringsAsFactors = F)
  rownames(pd) <- pd$id

  # gene expression
  ge <- read.delim("Verhaak.tsv", header = F, skip = 2, row.names = 1, colClasses = c(NA, "NULL", rep("numeric", nrow(pd))))
  colnames(ge) <- pd$id
  ge <- data.matrix(ge)

  # feature data
  fd <- select(org.Hs.eg.db::org.Hs.eg.db, rownames(ge), keys = rownames(ge), keytype = "ALIAS", columns = c("ENTREZID", "SYMBOL", "GENENAME"))

  # pick up gene symbols/aliases that can be translated to entrez id unambigiously
  da <- table(fd$ALIAS)
  da <- names(da[da == 1]) # disambigous aliases
  de <- table(fd$ENTREZID)
  de <- names(de[de == 1]) # disambigous entrez ids
  fd <- fd[fd$ALIAS %in% da & fd$ENTREZID %in% de,]

  fd$HUMENTREZID <- fd$ENTREZID
  rownames(fd) <- fd$HUMENTREZID

  # limit gene expression to new gene set
  all(fd$ALIAS %in% rownames(ge))
  ge <- ge[fd$ALIAS,]
  rownames(ge) <- rownames(fd)

  # create eSet
  V <- ExpressionSet(ge, AnnotatedDataFrame(pd), AnnotatedDataFrame(fd))
  V
}

V <- read_ver("Verhaak.tsv")

pca(V, lbs = c())

```


## Centroids

```{r}
get_ces <- function(V, ces_file){
  sel <- sapply(levels(pData(V)$class), function (x){
    which(pData(V)$class == x)[1]
  })

  V_ce <- V[,sel]
  colnames(V_ce) <- levels(pData(V)$class)
  pData(V_ce) <- pData(V_ce)[, colnames(pData(V_ce)) != "id", drop = F]

  ces <- gdata::read.xls(ces_file, header = T, row.names = 1,  stringsAsFactors = F, skip = 2, colClasses = c(NA, "NULL", rep(NA, 4)))
  ces <- ces[fData(V)$ALIAS, ]
  rownames(ces) <- rownames(fData(V))
  ces <- ces[,colnames(V_ce)]
  exprs(V_ce) <- data.matrix(ces)

  V_ce
}

V_ce <- get_ces(V, "ClaNC840_centroids.xls")

```


# Clustering

## Data modification

```{r}
get_DE <- function(R, ctrl = "CTRL"){
  base <- rowMeans(exprs(R)[, pData(R)$class == ctrl])
  exprs(R) <- sweep(exprs(R), 1, base)
  R#[, pData(R)$class != ctrl]
}

z_score <- function(R){
  exprs(R) <- sweep(exprs(R), 1, rowMeans(exprs(R)))
  exprs(R) <- sweep(exprs(R), 1 , apply(exprs(R), 1, sd), "/")
  R
}

med_norm <-  function(R){
  exprs(R) <- sweep(exprs(R), 1 , apply(exprs(R), 1, median))
  R
}


Rat_DE <- get_DE(Rat, ctrl = "Rat_Ctrl")
Mouse_DE <- get_DE(Mouse, ctrl = "Mouse_Ctrl")
G_DE <- get_DE(G, ctrl = "Control")

```


## Clustering and heatmap function

```{r}
clust <- function(..., fun = function(M) dist(t(M)), method = "complete", sel = rownames(eData[[1]])){
  eData <- list(...)

  sel <- Reduce(intersect, lapply(eData, rownames), sel)
  eData <- lapply(eData, function(x){x[sel,]})
  M <- do.call(cbind, lapply(eData, exprs))

  d <- fun(M)

  hc <- hclust(d, method = method)
  plot(hc, labels = unlist(sapply(eData, function(x){pData(x)$class})))

  invisible(hc)
}

heatm <- function(..., fun = function(M) dist(t(M)), method = "complete", sel = rownames(eData[[1]])){
  eData <- list(...)

  sel <- Reduce(intersect, lapply(eData, rownames), sel)
  eData <- lapply(eData, function(x){x[sel,]})
  M <- do.call(cbind, lapply(eData, exprs))

  cols <- colorRampPalette(c("#40FF40", "#FF0404"))(10)

  class <- unlist(sapply(eData, function(x){pData(x)$class}))
  hm <- heatmap(M, Rowv = NULL, Colv = NULL, labRow = c(""), labCol = c(""),
                distfun = function(x) fun(t(x)), hclustfun = function(d) hclust(d, method = method),
                col = cols, ColSideColors = palette()[as.numeric(class) + 1])
  col_key <- levels(class)
  legend("topleft", col_key, col = palette()[seq_along(col_key) + 1], lty = c(-1, -1), pch = c(19, 19))

  invisible(hm)
}


```


## Clustering results

### With GEO data

```{r}
clust(G_DE, Rat_DE, fun = function(M) as.dist(1-cor(M, method = "spearman")))
clust(G_DE, Mouse_DE, fun = function(M) as.dist(1-cor(M, method = "spearman")))
clust(G_DE, Rat_DE, Mouse_DE, fun = function(M) as.dist(1-cor(M, method = "spearman")))
```


### With Verhaak data

```{r}
clust(V_ce, med_norm(Rat), med_norm(Mouse), fun = function(M) as.dist(1-cor(M)), method = "complete")

clust(V_ce, med_norm(Rat), fun = function(M) as.dist(1-cor(M)), method = "complete")

clust(V_ce, med_norm(Mouse), fun = function(M) as.dist(1-cor(M)), method = "complete")


clust(V_ce, med_norm(Rat), med_norm(Mouse), fun = function(M) as.dist(1-cor(M)), method = "mcquitty")

clust(V_ce, med_norm(Rat), fun = function(M) as.dist(1-cor(M)), method = "mcquitty")

clust(V_ce, med_norm(Mouse), fun = function(M) as.dist(1-cor(M)), method = "mcquitty")


clust(V_ce, med_norm(Rat), med_norm(Mouse), fun = function(M) as.dist(1-cor(M)), method = "average")

clust(V_ce, med_norm(Rat), fun = function(M) as.dist(1-cor(M)), method = "average")

clust(V_ce, med_norm(Mouse), fun = function(M) as.dist(1-cor(M)), method = "average")

heatm(V, med_norm(Rat), med_norm(Mouse), fun = function(M) as.dist(1-cor(M)))

```

